gensim/similarity_learning/sl_vocab.py

```diff
@@ -0,0 +1,557 @@
+        ---------
+        batch_size: int
+            size of the batch"""
+        while True:
+            X1, X1_len, X2, X2_len, Y = self.get_batch(batch_size)
+            yield ({'query': X1, 'query_len': X1_len, 'doc': X2, 'doc_len': X2_len}, Y)
+
+
+class WikiQAExtractor:
+    """[WIP]Class to extract data from the WikiQA dataset
+    It provides data in a shape which makes it easier to train neural networks
+
+    Usage:
+    =====
+    wiki_extractor = WikiQAExtractor("path/to/file")
+    train_data = wiki_extractor.get_data()
+
+    the format provided is : [ [
+                                [query1, doc1, label_1_1],
+                                [query1, doc2, label_1_2],
+                                .
+                                .
+                               ],
+                               [
+                                [query2, doc1, label_2_1],
+                                [query2, doc2, label_2_2],
+                                .
+                                .
+                                ],
+                                .
+                                .
+                                .
+                               [
+                                [query_n, doc1, label_n_1],
+                                [query_n, doc2, label_n_2],
+                               ]
+                             ]
+    """
+
+    def __init__(self, file_path, word_embedding_path=None, embedding_dim=None, maxlen=30, triletter=False):
+        """
+        Parameters:
+        -----------
+        file_path: str
+            has to be a path to the tsv
+        """
+        if file_path is not None:
+            with open(file_path) as f:
+                self.df = pd.read_csv(f, sep='\t')
+        else:
+            raise NotImplementedError()
+
+        # TODO add 300k vector for all permutes
+        # TODO add option for using word embeddings
+
+        self.queries = list(self.df['Question'])
+        self.documents = list(self.df['Sentence'])
+        self.relations = list(self.df['Label'])
+        self.relations = [int(r) for r in self.relations]
+
+        self.triletter = triletter
+        self.embedding_dim = embedding_dim
+        self.maxlen = maxlen
+        self.word_embedding_path = word_embedding_path
+
+        self.word2index, self.index2word = {}, {}
+        self.word_counter = Counter()
+        self.corpus = self.queries + self.documents
+
+        if self.triletter:
+            self.tri2index, self.index2tri = {}, {}
+            self.triletter_counter = Counter()
+            self.triletter_corpus = []
+
+        self.build_vocab()
+
+    def preprocess(self, sentence):
+        return re.sub("[^a-zA-Z0-9]", " ", sentence.lower())
+
+    def get_preprocessed_corpus(self):
+        preprocessed_corpus = []
+        for sent in self.corpus:
+            preprocessed_corpus.append(self.preprocess(sent))
+        return preprocessed_corpus
+
+    def build_vocab(self):
```

Comment:
See current code organization for *2vec model please and make similar architecture here.